File: Data Structures & Algorithms/minimum-window-with-characters/submission-0.js
class Solution {
    /**
     * @param {string} s
     * @param {string} t
     * @return {string}
     */
    minWindow(s, t) {
        let l = 0, r = 0;
        let tMap = new Map();
        for (const c of t) {
            tMap.set(c, (tMap.get(c) || 0) + 1);
        }
        let sMap = new Map();
        sMap.set(s[l], 1);
        if (this.checkFreq(sMap, tMap)) return s[l];
        let ans = "";
        while (r < s.length) {
            if (this.checkFreq(sMap, tMap)) {
                if (ans.length > r - l + 1 || ans.length == 0)
                    ans = s.substring(l, r + 1);

                sMap.set(s[l], sMap.get(s[l]) - 1);
                l++;
            } else {
                r++;
                sMap.set(s[r], (sMap.get(s[r]) || 0) + 1);
            }
        }
        return ans;
    }
    checkFreq(sMap, tMap) {
        for (const [k, v] of tMap) {
            if ((sMap.get(k) || 0) < v) return false;
        }
        return true;
    };
}

File: Data Structures & Algorithms/is-palindrome/submission-0.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        for (char& c : s)
            c = tolower(c);
        string sf = clean(s);
        int i = 0, j = sf.size()-1;
        while(i<j && j>=0){
            if(sf[i]!=sf[j]){
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
    string clean(string s) {
        string ans = "";
        for (auto c : s) {
            if ((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9')) {
                ans += c;
            }
        }
        return ans;
    }
};
